File: docs/beginner/module-06-filters-sorting/code-samples/csharp/08_PerformanceAnalysis.cs
/*
 * Performance Analysis Example
 * 
 * This example demonstrates performance monitoring and optimization techniques
 * for Azure AI Search filter and sort operations.
 */

using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;
using Microsoft.Extensions.Configuration;
using System.Diagnostics;
using System.Text.Json;

namespace AzureSearchFiltersExamples
{
    public class PerformanceAnalysisExample
    {
        private readonly SearchClient _searchClient;
        private readonly IConfiguration _configuration;
        private readonly List<PerformanceMetric> _performanceHistory;

        public PerformanceAnalysisExample()
        {
            ValidateConfiguration();
            
            _configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json")
                .Build();

            var endpoint = new Uri(_configuration["SearchService:Endpoint"]);
            var apiKey = _configuration["SearchService:ApiKey"];
            var indexName = _configuration["SearchService:IndexName"];

            var credential = new AzureKeyCredential(apiKey);
            _searchClient = new SearchClient(endpoint, indexName, credential);
            _performanceHistory = new List<PerformanceMetric>();
        }

        private void ValidateConfiguration()
        {
            var config = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json")
                .Build();

            var endpoint = config["SearchService:Endpoint"];
            var apiKey = config["SearchService:ApiKey"];
            var indexName = config["SearchService:IndexName"];

            if (string.IsNullOrEmpty(endpoint) || string.IsNullOrEmpty(apiKey) || string.IsNullOrEmpty(indexName))
            {
                throw new InvalidOperationException("Missing required configuration. Check your appsettings.json file.");
            }

            Console.WriteLine("✅ Configuration validated");
            Console.WriteLine($"📍 Search Endpoint: {endpoint}");
            Console.WriteLine($"📊 Index Name: {indexName}");
        }

        public async Task<PerformanceMetric> MeasureQueryPerformanceAsync(string queryType, string searchText = "*", 
            string filter = null, string orderBy = null, int size = 10)
        {
            var stopwatch = Stopwatch.StartNew();
            var startMemory = GC.GetTotalMemory(false);

            try
            {
                var searchOptions = new SearchOptions
                {
                    Size = size,
                    IncludeTotalCount = true
                };

                if (!string.IsNullOrEmpty(filter))
                    searchOptions.Filter = filter;

                if (!string.IsNullOrEmpty(orderBy))
                    searchOptions.OrderBy.Add(orderBy);

                searchOptions.Select.Add("id");
                searchOptions.Select.Add("name");
                searchOptions.Select.Add("category");
                searchOptions.Select.Add("price");
                searchOptions.Select.Add("rating");

                var results = await _searchClient.SearchAsync<SearchDocument>(searchText, searchOptions);
                var resultList = new List<SearchResult<SearchDocument>>();

                await foreach (var result in results.Value.GetResultsAsync())
                {
                    resultList.Add(result);
                }

                stopwatch.Stop();
                var endMemory = GC.GetTotalMemory(false);
                var totalCount = results.Value.TotalCount ?? resultList.Count;

                var metric = new PerformanceMetric
                {
                    Timestamp = DateTime.UtcNow,
                    QueryType = queryType,
                    SearchText = searchText,
                    FilterExpression = filter ?? "none",
                    OrderBy = orderBy ?? "none",
                    ExecutionTimeMs = stopwatch.ElapsedMilliseconds,
                    ResultCount = totalCount,
                    ReturnedCount = resultList.Count,
                    MemoryUsageBytes = endMemory - startMemory,
                    ComplexityScore = CalculateComplexityScore(filter),
                    Success = true
                };

                _performanceHistory.Add(metric);
                return metric;
            }
            catch (Exception ex)
            {
                stopwatch.Stop();
                var endMemory = GC.GetTotalMemory(false);

                var metric = new PerformanceMetric
                {
                    Timestamp = DateTime.UtcNow,
                    QueryType = queryType,
                    SearchText = searchText,
                    FilterExpression = filter ?? "none",
                    OrderBy = orderBy ?? "none",
                    ExecutionTimeMs = stopwatch.ElapsedMilliseconds,
                    MemoryUsageBytes = endMemory - startMemory,
                    ComplexityScore = CalculateComplexityScore(filter),
                    Success = false,
                    ErrorMessage = ex.Message
                };

                _performanceHistory.Add(metric);
                return metric;
            }
        }

        public async Task RunPerformanceBenchmarkAsync()
        {
            Console.WriteLine("\n🚀 Performance Benchmark Suite");
            Console.WriteLine("=".PadRight(40, '='));

            var testScenarios = new[]
            {
                new TestScenario
                {
                    Name = "Baseline Search",
                    SearchText = "*",
                    Filter = null,
                    OrderBy = null,
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Simple Filter",
                    SearchText = "*",
                    Filter = "rating ge 4.0",
                    OrderBy = null,
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Range Filter",
                    SearchText = "*",
                    Filter = "price ge 50 and price le 200",
                    OrderBy = null,
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Collection Filter",
                    SearchText = "*",
                    Filter = "tags/any(tag: tag eq 'premium')",
                    OrderBy = null,
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Complex Filter",
                    SearchText = "*",
                    Filter = "(tags/any(tag: tag eq 'premium') or rating gt 4.5) and price le 300",
                    OrderBy = null,
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Geographic Filter",
                    SearchText = "*",
                    Filter = "geo.distance(location, geography'POINT(-122.3321 47.6062)') lt 10",
                    OrderBy = null,
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Sorted Results",
                    SearchText = "*",
                    Filter = null,
                    OrderBy = "rating desc",
                    Iterations = 5
                },
                new TestScenario
                {
                    Name = "Complex Sort",
                    SearchText = "*",
                    Filter = null,
                    OrderBy = "rating desc, price asc",
                    Iterations = 5
                }
            };

            var benchmarkResults = new List<BenchmarkResult>();

            foreach (var scenario in testScenarios)
            {
                Console.WriteLine($"\n📊 Testing: {scenario.Name}");
                Console.WriteLine($"   Filter: {scenario.Filter ?? "none"}");
                Console.WriteLine($"   Order: {scenario.OrderBy ?? "none"}");
                Console.WriteLine($"   Iterations: {scenario.Iterations}");

                var metrics = new List<PerformanceMetric>();

                for (int i = 0; i < scenario.Iterations; i++)
                {
                    var metric = await MeasureQueryPerformanceAsync(
                        scenario.Name, scenario.SearchText, scenario.Filter, scenario.OrderBy);
                    
                    if (metric.Success)
                        metrics.Add(metric);

                    // Small delay between iterations
                    await Task.Delay(100);
                }

                if (metrics.Any())
                {
                    var result = new BenchmarkResult
                    {
                        ScenarioName = scenario.Name,
                        SuccessfulRuns = metrics.Count,
                        TotalRuns = scenario.Iterations,
                        AvgExecutionTimeMs = metrics.Average(m => m.ExecutionTimeMs),
                        MinExecutionTimeMs = metrics.Min(m => m.ExecutionTimeMs),
                        MaxExecutionTimeMs = metrics.Max(m => m.ExecutionTimeMs),
                        StdDevExecutionTimeMs = CalculateStandardDeviation(metrics.Select(m => (double)m.ExecutionTimeMs)),
                        AvgResultCount = metrics.Average(m => m.ResultCount),
                        AvgMemoryUsageBytes = metrics.Average(m => m.MemoryUsageBytes),
                        ComplexityScore = metrics.First().ComplexityScore,
                        Filter = scenario.Filter
                    };

                    benchmarkResults.Add(result);

                    Console.WriteLine($"   ✅ Avg Time: {result.AvgExecutionTimeMs:F1}ms");
                    Console.WriteLine($"   📊 Results: {result.AvgResultCount:F0}");
                    Console.WriteLine($"   🎯 Complexity: {result.ComplexityScore}");
                }
                else
                {
                    Console.WriteLine($"   ❌ All iterations failed");
                }
            }

            // Generate benchmark report
            GenerateBenchmarkReport(benchmarkResults);
        }

        public void GenerateBenchmarkReport(List<BenchmarkResult> results)
        {
            Console.WriteLine("\n📊 BENCHMARK REPORT");
            Console.WriteLine("=".PadRight(50, '='));

            if (!results.Any())
            {
                Console.WriteLine("No successful benchmark results to report.");
                return;
            }

            // Summary statistics
            var avgTime = results.Average(r => r.AvgExecutionTimeMs);
            var fastestScenario = results.OrderBy(r => r.AvgExecutionTimeMs).First();
            var slowestScenario = results.OrderByDescending(r => r.AvgExecutionTimeMs).First();

            Console.WriteLine($"📈 Summary Statistics:");
            Console.WriteLine($"   Total Scenarios: {results.Count}");
            Console.WriteLine($"   Average Execution Time: {avgTime:F1}ms");
            Console.WriteLine($"   Fastest Scenario: {fastestScenario.ScenarioName} ({fastestScenario.AvgExecutionTimeMs:F1}ms)");
            Console.WriteLine($"   Slowest Scenario: {slowestScenario.ScenarioName} ({slowestScenario.AvgExecutionTimeMs:F1}ms)");

            // Detailed results
            Console.WriteLine($"\n🔍 Detailed Results:");
            Console.WriteLine($"{"Scenario",-25} {"Avg Time",-10} {"Results",-8} {"Complexity",-10} {"Performance",-12}");
            Console.WriteLine(new string('-', 70));

            foreach (var result in results.OrderBy(r => r.AvgExecutionTimeMs))
            {
                var performance = result.AvgExecutionTimeMs < 100 ? "Excellent" :
                                result.AvgExecutionTimeMs < 500 ? "Good" :
                                result.AvgExecutionTimeMs < 1000 ? "Fair" : "Needs work";

                Console.WriteLine($"{result.ScenarioName,-25} {result.AvgExecutionTimeMs,-10:F1}ms {result.AvgResultCount,-8:F0} {result.ComplexityScore,-10} {performance,-12}");
            }

            // Performance recommendations
            GenerateOptimizationRecommendations(results);
        }

        public void GenerateOptimizationRecommendations(List<BenchmarkResult> results)
        {
            Console.WriteLine($"\n💡 Optimization Recommendations:");
            Console.WriteLine(new string('-', 35));

            var recommendations = new List<string>();

            // Analyze slow queries
            var slowQueries = results.Where(r => r.AvgExecutionTimeMs > 1000).ToList();
            if (slowQueries.Any())
            {
                recommendations.Add($"🔴 HIGH PRIORITY: {slowQueries.Count} slow queries (>1s) need optimization");
                foreach (var slow in slowQueries)
                {
                    recommendations.Add($"   • {slow.ScenarioName}: {slow.AvgExecutionTimeMs:F1}ms");
                }
            }

            // Analyze complexity
            var complexQueries = results.Where(r => r.ComplexityScore > 15).ToList();
            if (complexQueries.Any())
            {
                recommendations.Add($"🟡 MEDIUM PRIORITY: {complexQueries.Count} high-complexity filters detected");
                recommendations.Add("   • Consider simplifying filter expressions");
                recommendations.Add("   • Use faceted navigation for complex OR conditions");
            }

            // Analyze geographic queries
            var geoQueries = results.Where(r => r.Filter?.Contains("geo.distance") == true).ToList();
            if (geoQueries.Any())
            {
                recommendations.Add($"🔵 INFO: {geoQueries.Count} geographic queries detected");
                recommendations.Add("   • Consider using smaller search radii");
                recommendations.Add("   • Ensure geographic fields are properly indexed");
            }

            // General recommendations
            recommendations.Add("🟢 GENERAL RECOMMENDATIONS:");
            recommendations.Add("   • Mark frequently filtered fields as 'filterable' in index schema");
            recommendations.Add("   • Use appropriate data types for optimal performance");
            recommendations.Add("   • Consider caching results for frequently used queries");
            recommendations.Add("   • Monitor query performance in production");

            foreach (var recommendation in recommendations)
            {
                Console.WriteLine(recommendation);
            }
        }

        public void AnalyzePerformanceTrends()
        {
            Console.WriteLine("\n📈 Performance Trend Analysis");
            Console.WriteLine("=".PadRight(40, '='));

            if (_performanceHistory.Count < 5)
            {
                Console.WriteLine("Insufficient data for trend analysis (need at least 5 measurements)");
                return;
            }

            // Group by query type
            var groupedMetrics = _performanceHistory
                .Where(m => m.Success)
                .GroupBy(m => m.QueryType)
                .ToList();

            foreach (var group in groupedMetrics)
            {
                var metrics = group.OrderBy(m => m.Timestamp).ToList();
                if (metrics.Count < 3) continue;

                var avgTime = metrics.Average(m => m.ExecutionTimeMs);
                var trend = CalculateTrend(metrics.Select(m => (double)m.ExecutionTimeMs).ToList());

                Console.WriteLine($"\n📊 {group.Key}:");
                Console.WriteLine($"   Measurements: {metrics.Count}");
                Console.WriteLine($"   Average Time: {avgTime:F1}ms");
                Console.WriteLine($"   Trend: {(trend > 0.1 ? "⬆️ Degrading" : trend < -0.1 ? "⬇️ Improving" : "➡️ Stable")}");
                Console.WriteLine($"   Latest: {metrics.Last().ExecutionTimeMs}ms");
            }
        }

        public void ExportPerformanceData(string filename = null)
        {
            filename ??= $"performance_data_{DateTime.Now:yyyyMMdd_HHmmss}.json";

            try
            {
                var exportData = new
                {
                    ExportTimestamp = DateTime.UtcNow,
                    TotalMeasurements = _performanceHistory.Count,
                    SuccessfulMeasurements = _performanceHistory.Count(m => m.Success),
                    DateRange = new
                    {
                        Start = _performanceHistory.Any() ? _performanceHistory.Min(m => m.Timestamp) : (DateTime?)null,
                        End = _performanceHistory.Any() ? _performanceHistory.Max(m => m.Timestamp) : (DateTime?)null
                    },
                    Metrics = _performanceHistory
                };

                var json = JsonSerializer.Serialize(exportData, new JsonSerializerOptions 
                { 
                    WriteIndented = true,
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                });

                File.WriteAllText(filename, json);
                Console.WriteLine($"✅ Performance data exported to {filename}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Export failed: {ex.Message}");
            }
        }

        private int CalculateComplexityScore(string filter)
        {
            if (string.IsNullOrEmpty(filter))
                return 0;

            var score = 0;
            score += CountOccurrences(filter, " and ") * 1;
            score += CountOccurrences(filter, " or ") * 2;
            score += CountOccurrences(filter, "/any(") * 3;
            score += CountOccurrences(filter, "/all(") * 4;
            score += CountOccurrences(filter, "geo.distance") * 3;
            score += CountOccurrences(filter, "(") * 1;

            return score;
        }

        private int CountOccurrences(string text, string pattern)
        {
            return (text.Length - text.Replace(pattern, "").Length) / pattern.Length;
        }

        private double CalculateStandardDeviation(IEnumerable<double> values)
        {
            var valuesList = values.ToList();
            if (valuesList.Count <= 1) return 0;

            var average = valuesList.Average();
            var sumOfSquaresOfDifferences = valuesList.Select(val => (val - average) * (val - average)).Sum();
            return Math.Sqrt(sumOfSquaresOfDifferences / (valuesList.Count - 1));
        }

        private double CalculateTrend(List<double> values)
        {
            if (values.Count < 2) return 0;

            var n = values.Count;
            var sumX = Enumerable.Range(0, n).Sum();
            var sumY = values.Sum();
            var sumXY = values.Select((y, x) => x * y).Sum();
            var sumXX = Enumerable.Range(0, n).Select(x => x * x).Sum();

            return (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);
        }

        public async Task RunAsync()
        {
            Console.WriteLine("🚀 Performance Analysis Example");
            Console.WriteLine("=".PadRight(50, '='));

            try
            {
                await RunPerformanceBenchmarkAsync();
                AnalyzePerformanceTrends();
                ExportPerformanceData();

                Console.WriteLine("\n✅ Performance analysis completed successfully!");
                Console.WriteLine($"\n📊 Total measurements collected: {_performanceHistory.Count}");
                Console.WriteLine($"✅ Successful queries: {_performanceHistory.Count(m => m.Success)}");
                Console.WriteLine($"❌ Failed queries: {_performanceHistory.Count(m => !m.Success)}");
                
                Console.WriteLine("\nKey takeaways:");
                Console.WriteLine("- Monitor query performance regularly");
                Console.WriteLine("- Optimize slow and complex queries");
                Console.WriteLine("- Use appropriate indexing strategies");
                Console.WriteLine("- Consider caching for frequently used queries");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\n❌ Example failed: {ex.Message}");
                throw;
            }
        }
    }

    public class PerformanceMetric
    {
        public DateTime Timestamp { get; set; }
        public string QueryType { get; set; }
        public string SearchText { get; set; }
        public string FilterExpression { get; set; }
        public string OrderBy { get; set; }
        public long ExecutionTimeMs { get; set; }
        public int ResultCount { get; set; }
        public int ReturnedCount { get; set; }
        public long MemoryUsageBytes { get; set; }
        public int ComplexityScore { get; set; }
        public bool Success { get; set; }
        public string ErrorMessage { get; set; }
    }

    public class TestScenario
    {
        public string Name { get; set; }
        public string SearchText { get; set; }
        public string Filter { get; set; }
        public string OrderBy { get; set; }
        public int Iterations { get; set; }
    }

    public class BenchmarkResult
    {
        public string ScenarioName { get; set; }
        public int SuccessfulRuns { get; set; }
        public int TotalRuns { get; set; }
        public double AvgExecutionTimeMs { get; set; }
        public long MinExecutionTimeMs { get; set; }
        public long MaxExecutionTimeMs { get; set; }
        public double StdDevExecutionTimeMs { get; set; }
        public double AvgResultCount { get; set; }
        public double AvgMemoryUsageBytes { get; set; }
        public int ComplexityScore { get; set; }
        public string Filter { get; set; }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            var example = new PerformanceAnalysisExample();
            try
            {
                await example.RunAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Application failed: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: docs/beginner/module-06-filters-sorting/code-samples/python/08_performance_analysis.py
#!/usr/bin/env python3
"""
Azure AI Search - Performance Analysis Dashboard

This script provides comprehensive performance analysis for Azure AI Search
filter and sort operations, including real-time monitoring, optimization
strategies, and scalability considerations.

Key Features:
- Real-time query performance monitoring
- Filter optimization strategies
- Resource usage pattern analysis
- Optimization recommendations
- Scalability planning
- Performance bottleneck identification
- Comparative analysis of filtering approaches

Prerequisites:
- Azure AI Search service configured
- Sample data loaded in your index
- Environment variables set in .env file
"""

import os
import time
import statistics
import json
from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass, asdict
from datetime import datetime, timedelta
from azure.search.documents import SearchClient
from azure.core.credentials import AzureKeyCredential
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

@dataclass
class PerformanceMetric:
    """Represents a performance measurement."""
    timestamp: datetime
    query_type: str
    filter_expression: str
    execution_time: float
    result_count: int
    complexity_score: int
    memory_usage: Optional[float] = None
    cpu_usage: Optional[float] = None

@dataclass
class OptimizationRecommendation:
    """Represents an optimization recommendation."""
    category: str
    priority: str  # 'high', 'medium', 'low'
    description: str
    impact: str
    implementation: str

class PerformanceAnalyzer:
    """Analyzes and monitors Azure AI Search performance."""
    
    def __init__(self):
        """Initialize the performance analyzer."""
        self.search_client = self._initialize_client()
        self.performance_history: List[PerformanceMetric] = []
        self.baseline_metrics = {}
        
    def _initialize_client(self) -> SearchClient:
        """Initialize Azure AI Search client."""
        try:
            endpoint = os.getenv('SEARCH_ENDPOINT')
            api_key = os.getenv('SEARCH_API_KEY')
            index_name = os.getenv('INDEX_NAME')
            
            if not all([endpoint, api_key, index_name]):
                raise ValueError("Missing required environment variables")
            
            credential = AzureKeyCredential(api_key)
            client = SearchClient(endpoint, index_name, credential)
            
            print(f"✅ Connected to Azure AI Search")
            print(f"📍 Endpoint: {endpoint}")
            print(f"📊 Index: {index_name}")
            
            return client
            
        except Exception as e:
            print(f"❌ Configuration error: {e}")
            raise
    
    def calculate_complexity_score(self, filter_expr: str) -> int:
        """Calculate complexity score for a filter expression."""
        if not filter_expr or filter_expr == "*":
            return 0
        
        score = 0
        score += filter_expr.count('and') * 1
        score += filter_expr.count('or') * 2
        score += filter_expr.count('/any') * 3
        score += filter_expr.count('/all') * 4
        score += filter_expr.count('geo.distance') * 3
        score += filter_expr.count('(') * 1
        score += len(filter_expr.split()) * 0.1
        
        return int(score)
    
    def measure_query_performance(self, search_text: str = "*", 
                                filter_expr: str = None,
                                orderby: str = None,
                                top: int = 10,
                                query_type: str = "standard") -> PerformanceMetric:
        """
        Measure performance of a single query.
        
        Args:
            search_text: Search query text
            filter_expr: Filter expression
            orderby: Sort expression
            top: Number of results
            query_type: Type of query for categorization
            
        Returns:
            Performance metric
        """
        try:
            start_time = time.time()
            
            search_params = {
                'search_text': search_text,
                'top': top,
                'include_total_count': True
            }
            
            if filter_expr:
                search_params['filter'] = filter_expr
            if orderby:
                search_params['order_by'] = orderby
            
            results = self.search_client.search(**search_params)
            
            # Process results to get count
            documents = list(results)
            result_count = results.get_count() or len(documents)
            
            execution_time = time.time() - start_time
            complexity_score = self.calculate_complexity_score(filter_expr or "")
            
            metric = PerformanceMetric(
                timestamp=datetime.now(),
                query_type=query_type,
                filter_expression=filter_expr or "none",
                execution_time=execution_time,
                result_count=result_count,
                complexity_score=complexity_score
            )
            
            self.performance_history.append(metric)
            return metric
            
        except Exception as e:
            print(f"❌ Performance measurement error: {e}")
            return None
    
    def run_performance_benchmark(self, test_scenarios: List[Dict] = None) -> Dict:
        """
        Run a comprehensive performance benchmark.
        
        Args:
            test_scenarios: List of test scenarios to run
            
        Returns:
            Benchmark results
        """
        if test_scenarios is None:
            test_scenarios = self._get_default_test_scenarios()
        
        print("🚀 Running Performance Benchmark...")
        print("=" * 40)
        
        benchmark_results = {
            'scenarios': [],
            'summary': {},
            'recommendations': []
        }
        
        for scenario in test_scenarios:
            print(f"\n📊 Testing: {scenario['name']}")
            
            # Run multiple iterations for statistical accuracy
            iterations = scenario.get('iterations', 3)
            metrics = []
            
            for i in range(iterations):
                metric = self.measure_query_performance(
                    search_text=scenario.get('search_text', '*'),
                    filter_expr=scenario.get('filter'),
                    orderby=scenario.get('orderby'),
                    top=scenario.get('top', 10),
                    query_type=scenario['name']
                )
                
                if metric:
                    metrics.append(metric)
                
                # Small delay between iterations
                time.sleep(0.1)
            
            if metrics:
                # Calculate statistics
                execution_times = [m.execution_time for m in metrics]
                result_counts = [m.result_count for m in metrics]
                
                scenario_result = {
                    'name': scenario['name'],
                    'avg_execution_time': statistics.mean(execution_times),
                    'min_execution_time': min(execution_times),
                    'max_execution_time': max(execution_times),
                    'std_execution_time': statistics.stdev(execution_times) if len(execution_times) > 1 else 0,
                    'avg_result_count': statistics.mean(result_counts),
                    'complexity_score': metrics[0].complexity_score,
                    'filter': scenario.get('filter', 'none')
                }
                
                benchmark_results['scenarios'].append(scenario_result)
                
                print(f"   Avg Time: {scenario_result['avg_execution_time']:.3f}s")
                print(f"   Results: {scenario_result['avg_result_count']:.0f}")
                print(f"   Complexity: {scenario_result['complexity_score']}")
        
        # Generate summary and recommendations
        benchmark_results['summary'] = self._generate_benchmark_summary(benchmark_results['scenarios'])
        benchmark_results['recommendations'] = self._generate_optimization_recommendations(benchmark_results['scenarios'])
        
        return benchmark_results
    
    def _get_default_test_scenarios(self) -> List[Dict]:
        """Get default test scenarios for benchmarking."""
        return [
            {
                'name': 'Baseline Search',
                'search_text': '*',
                'iterations': 5
            },
            {
                'name': 'Simple Filter',
                'search_text': '*',
                'filter': "rating ge 4.0",
                'iterations': 5
            },
            {
                'name': 'Range Filter',
                'search_text': '*',
                'filter': "price ge 50 and price le 200",
                'iterations': 5
            },
            {
                'name': 'Collection Filter',
                'search_text': '*',
                'filter': "tags/any(item: item eq 'premium')",
                'iterations': 5
            },
            {
                'name': 'Complex Filter',
                'search_text': '*',
                'filter': "(tags/any(item: item eq 'premium') or rating gt 4.5) and price le 300",
                'iterations': 5
            },
            {
                'name': 'Geographic Filter',
                'search_text': '*',
                'filter': "geo.distance(location, geography'POINT(-122.3321 47.6062)') lt 10",
                'iterations': 5
            },
            {
                'name': 'Sorted Results',
                'search_text': '*',
                'orderby': 'rating desc',
                'iterations': 5
            },
            {
                'name': 'Complex Sort',
                'search_text': '*',
                'orderby': 'rating desc, price asc',
                'iterations': 5
            }
        ]
    
    def _generate_benchmark_summary(self, scenarios: List[Dict]) -> Dict:
        """Generate summary statistics from benchmark results."""
        if not scenarios:
            return {}
        
        execution_times = [s['avg_execution_time'] for s in scenarios]
        complexity_scores = [s['complexity_score'] for s in scenarios]
        
        return {
            'total_scenarios': len(scenarios),
            'avg_execution_time': statistics.mean(execution_times),
            'fastest_scenario': min(scenarios, key=lambda x: x['avg_execution_time'])['name'],
            'slowest_scenario': max(scenarios, key=lambda x: x['avg_execution_time'])['name'],
            'avg_complexity': statistics.mean(complexity_scores),
            'performance_variance': statistics.stdev(execution_times) if len(execution_times) > 1 else 0
        }
    
    def _generate_optimization_recommendations(self, scenarios: List[Dict]) -> List[OptimizationRecommendation]:
        """Generate optimization recommendations based on benchmark results."""
        recommendations = []
        
        # Analyze slow queries
        slow_threshold = 1.0  # 1 second
        slow_scenarios = [s for s in scenarios if s['avg_execution_time'] > slow_threshold]
        
        if slow_scenarios:
            recommendations.append(OptimizationRecommendation(
                category="Query Performance",
                priority="high",
                description=f"Found {len(slow_scenarios)} slow queries (>{slow_threshold}s)",
                impact="Significant performance improvement",
                implementation="Review and optimize slow filter expressions"
            ))
        
        # Analyze complexity
        high_complexity = [s for s in scenarios if s['complexity_score'] > 15]
        if high_complexity:
            recommendations.append(OptimizationRecommendation(
                category="Filter Complexity",
                priority="medium",
                description=f"Found {len(high_complexity)} high-complexity filters",
                impact="Moderate performance improvement",
                implementation="Simplify complex filter expressions or use faceted navigation"
            ))
        
        # Check for geographic queries
        geo_scenarios = [s for s in scenarios if 'geo.distance' in s.get('filter', '')]
        if geo_scenarios:
            recommendations.append(OptimizationRecommendation(
                category="Geographic Queries",
                priority="medium",
                description="Geographic queries detected",
                impact="Location-based performance optimization",
                implementation="Consider using smaller search radii and geographic indexing"
            ))
        
        # General recommendations
        recommendations.append(OptimizationRecommendation(
            category="Index Configuration",
            priority="low",
            description="Ensure proper field indexing",
            impact="Overall performance improvement",
            implementation="Mark frequently filtered fields as 'filterable' in index schema"
        ))
        
        return recommendations
    
    def analyze_performance_trends(self, hours: int = 24) -> Dict:
        """
        Analyze performance trends over time.
        
        Args:
            hours: Number of hours to analyze
            
        Returns:
            Trend analysis results
        """
        cutoff_time = datetime.now() - timedelta(hours=hours)
        recent_metrics = [m for m in self.performance_history if m.timestamp >= cutoff_time]
        
        if not recent_metrics:
            return {'error': 'No recent performance data available'}
        
        # Group by query type
        by_type = {}
        for metric in recent_metrics:
            if metric.query_type not in by_type:
                by_type[metric.query_type] = []
            by_type[metric.query_type].append(metric)
        
        trends = {}
        for query_type, metrics in by_type.items():
            execution_times = [m.execution_time for m in metrics]
            trends[query_type] = {
                'count': len(metrics),
                'avg_time': statistics.mean(execution_times),
                'trend': 'stable'  # Simplified - would need more data for real trend analysis
            }
        
        return {
            'period_hours': hours,
            'total_queries': len(recent_metrics),
            'query_types': trends,
            'overall_avg_time': statistics.mean([m.execution_time for m in recent_metrics])
        }
    
    def generate_performance_report(self, benchmark_results: Dict = None) -> str:
        """Generate a comprehensive performance report."""
        report = []
        report.append("📊 Azure AI Search Performance Analysis Report")
        report.append("=" * 55)
        report.append(f"Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        report.append("")
        
        if benchmark_results:
            summary = benchmark_results.get('summary', {})
            scenarios = benchmark_results.get('scenarios', [])
            recommendations = benchmark_results.get('recommendations', [])
            
            # Summary section
            report.append("📈 Performance Summary")
            report.append("-" * 25)
            report.append(f"Total Scenarios Tested: {summary.get('total_scenarios', 0)}")
            report.append(f"Average Execution Time: {summary.get('avg_execution_time', 0):.3f}s")
            report.append(f"Fastest Scenario: {summary.get('fastest_scenario', 'N/A')}")
            report.append(f"Slowest Scenario: {summary.get('slowest_scenario', 'N/A')}")
            report.append(f"Performance Variance: {summary.get('performance_variance', 0):.3f}s")
            report.append("")
            
            # Detailed results
            report.append("🔍 Detailed Results")
            report.append("-" * 20)
            for scenario in scenarios:
                report.append(f"• {scenario['name']}")
                report.append(f"  Time: {scenario['avg_execution_time']:.3f}s ± {scenario['std_execution_time']:.3f}s")
                report.append(f"  Results: {scenario['avg_result_count']:.0f}")
                report.append(f"  Complexity: {scenario['complexity_score']}")
                if scenario['filter'] != 'none':
                    report.append(f"  Filter: {scenario['filter']}")
                report.append("")
            
            # Recommendations
            if recommendations:
                report.append("💡 Optimization Recommendations")
                report.append("-" * 35)
                for rec in recommendations:
                    report.append(f"• {rec.category} ({rec.priority.upper()} priority)")
                    report.append(f"  {rec.description}")
                    report.append(f"  Impact: {rec.impact}")
                    report.append(f"  Action: {rec.implementation}")
                    report.append("")
        
        # Historical trends
        trends = self.analyze_performance_trends()
        if 'error' not in trends:
            report.append("📊 Recent Performance Trends")
            report.append("-" * 30)
            report.append(f"Period: Last {trends['period_hours']} hours")
            report.append(f"Total Queries: {trends['total_queries']}")
            report.append(f"Overall Average Time: {trends['overall_avg_time']:.3f}s")
            report.append("")
            
            for query_type, data in trends['query_types'].items():
                report.append(f"• {query_type}: {data['count']} queries, {data['avg_time']:.3f}s avg")
        
        return "\n".join(report)
    
    def export_performance_data(self, filename: str = None) -> str:
        """Export performance data to JSON file."""
        if filename is None:
            filename = f"performance_data_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
        
        data = {
            'export_timestamp': datetime.now().isoformat(),
            'metrics': [asdict(metric) for metric in self.performance_history],
            'summary': {
                'total_queries': len(self.performance_history),
                'date_range': {
                    'start': min(m.timestamp for m in self.performance_history).isoformat() if self.performance_history else None,
                    'end': max(m.timestamp for m in self.performance_history).isoformat() if self.performance_history else None
                }
            }
        }
        
        # Convert datetime objects to strings for JSON serialization
        for metric in data['metrics']:
            metric['timestamp'] = metric['timestamp'].isoformat()
        
        try:
            with open(filename, 'w') as f:
                json.dump(data, f, indent=2)
            print(f"✅ Performance data exported to {filename}")
            return filename
        except Exception as e:
            print(f"❌ Export error: {e}")
            return None

def main():
    """Main demonstration function."""
    try:
        # Initialize the performance analyzer
        analyzer = PerformanceAnalyzer()
        
        print("🚀 Starting Performance Analysis...")
        print()
        
        # Run benchmark
        benchmark_results = analyzer.run_performance_benchmark()
        
        # Generate and display report
        print("\n" + "="*60)
        report = analyzer.generate_performance_report(benchmark_results)
        print(report)
        
        # Export data
        print("\n📁 Exporting Performance Data...")
        filename = analyzer.export_performance_data()
        
        print(f"\n✅ Performance analysis completed!")
        print(f"📊 Analyzed {len(benchmark_results['scenarios'])} scenarios")
        print(f"💡 Generated {len(benchmark_results['recommendations'])} recommendations")
        
    except Exception as e:
        print(f"❌ Error: {e}")

if __name__ == "__main__":
    main()